### PR TITLE
Upload for testing first. Will provide full ASAP

### PR DESCRIPTION
**Pull request by theanh28**

### CODE REVIEW

**Comment by theanh28** [comment]
Hi,
My name is Anh Vu. A friend of mine introduced me to the gold mine challenge and I hope I can use this opportunity to solve a great coding question, getting onto the leaderboard, and practicing js (and getting some sweet internship chance too, if possible). 

My first approach to the problem is an implementation of DP/BFS (they are somewhat alike, states going to other states). Basically, each position in the map (gold mine) can be reached in 3 ways: go down right, go right, and up right. We will try to calculate/optimize the answers to function a(position, reached by) and use them to optimize other func a().

I hope you can merge my pull request and see how far this solution may go.

Sorry for the inexperience pull request and JavaScript coding. It is my first time contributing to an open source and trying js

File: src/runner.js [hunk 1]
```diff
@@ -1,65 +1,127 @@
...
+
+/**
+ * Given a mine, runs the miner through the mine collecting gold along the way.
+ *
+ * @param  {array} mine - A n x m multidimensional array respresenting the mine.
+ * @param  {string} logFile - A file location where moves of the miner should be logged to.
+ * @param  {Number} yStart - The y dimension starting position for the miner.
+ *
+ * @return {Number} The total gold collected by the miner.
+ */
```
Comment (jordanell) [inline src/runner.js:14]: This comment block no longer matches the function below it. I would expect this to be updated to reflect the `optimize` function using the correct jsDoc syntax.

File: src/runner.js [hunk 2]
```diff
@@ -1,65 +1,127 @@
...
+
+function optimize(a, modeA, b, modeB, val) {
```
Comment (jordanell) [inline src/runner.js:16]: Using variables like `a`, `b` and even `modeA` etc make it very hard to tell what this function does by glimpsing at the signature. I would have to read the function's body to really know what all these variables mean.

File: src/runner.js [hunk 3]
```diff
@@ -1,65 +1,127 @@
...
+  if (a > b) return [a + val, modeA];
+  return [b + val, modeB];
+}
+function update(ans, val, x, y) {
```
Comment (jordanell) [inline src/runner.js:20]: I would like to see `jsDoc` style comments on any newly added functions.

File: src/runner.js [hunk 4]
```diff
@@ -1,65 +1,127 @@
...
+  if (ans.val > val) return;
+  ans.val = val;
+  ans.x = x;
+  ans.y = y;
```
Comment (jordanell) [inline src/runner.js:24]: Again, I would not use such terse variable names as it makes the code very hard to read / understand.

File: src/runner.js [hunk 5]
```diff
@@ -1,65 +1,127 @@
...
+  //---------------
+  /* print out a[] for debug
+  let p = [];
+  for (let i = 0; i < mine[0].length; ++i) {
+      p.push([]);
+      for (let j = 0; j < mine.length; ++j)
+          p[i].push(0);
+  }
+  */
+  //---------------
```
Comment (jordanell) [inline src/runner.js:39]: For final version of code, make sure you remove any commented out code that was used for debugging purposes.

File: src/runner.js [hunk 6]
```diff
@@ -1,65 +1,127 @@
...
+  let a = [ [ [], [], [] ] ]; //our BFS value board
+  let t = [ [ [], [], [] ] ]; //previous MOVE to backtrack from final position
```
Comment (jordanell) [inline src/runner.js:41]: Again with the terse variable names. Also, I wouldn't mind seeing a comment explaining this very odd variable initialization. Why do we have an array inside an array of 3 arrays?

File: src/runner.js [hunk 7]
```diff
@@ -1,65 +1,127 @@
...
+  for (let i = 0; i < mine.length; ++i) { //initial move
+      a[0][0].push(mine[i][0]); //filling array
+      a[0][1].push(mine[i][0]);
+      a[0][2].push(mine[i][0]);
+      t[0][0].push(-1); //filling backtrack
+      t[0][1].push(-1);
+      t[0][2].push(-1);
+  }
```
Comment (jordanell) [inline src/runner.js:49]: This might be better off in it's own function.

File: src/runner.js [hunk 8]
```diff
@@ -1,65 +1,127 @@
...
+      trace.push(`${ans.x},${ans.y}`);
+      let temp = ans.y;
+      ans.y -= (ans.m - 1); // 0 -> + 1, 1 -> 0, 2 -> -1
+      ans.m = t[ans.x][ans.m][temp];
+      -- ans.x;
+  }
+  for (let i = trace.length - 1; i >= 0; --i)
+      log(logFile, trace[i]);
+  return ans.val;
+}
```
Comment (jordanell) [inline src/runner.js:110]: This is pretty hard to read. In production ready code I would expect:

1. Better variable naming
2. More comments explaining what is happening
3. Probably breaking this down into several small functions

**Comment by jordanell** [comment]
If you are using this challenge for learning purposes, I would suggest you also look to try and add some unit testing to this package. Perhaps the `mocha` and `chai` NPM packages would be a good starting point :).

File: src/runner.js (hunk 1 above)
Comment (theanh28) [reply src/runner.js:14]: Absolutely. I'll fix it right away.

Comment (jordanell) [reply src/runner.js:14]: No need to really fix, I am just providing feedback for how I would review a PR in a live environment :).

**Pull request closed without merging**